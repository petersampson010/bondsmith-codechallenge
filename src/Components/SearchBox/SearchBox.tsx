import React, { useState, useRef, useEffect, useCallback } from 'react';
import { FC } from 'react';
import { useDispatch } from 'react-redux';
import { updateSearchShows } from '../../ReduxStore/actions.tsx';
import { getSearchedShows } from '../../Utils/api.tsx';
const _ = require('lodash');

const SearchBox: FC = () => {

    const dispatch = useDispatch(),
    [searchEntry, updateSearchEntry] = useState('');

    const handleDebounceFn = async(inputValue) => {
        window.history.replaceState({}, document.title, '?search=' + inputValue);
        const shows = await getSearchedShows(inputValue);
        console.log(shows);
        dispatch(updateSearchShows(shows));
    }
    
    const debounceFn = useCallback(_.debounce(handleDebounceFn, 400), []);

    const runSearch = (e) => updateSearchEntry(e.target.value);

    const handleChange = (e) => {
        updateSearchEntry(e.target.value);
        debounceFn(e.target.value);
    }

    return (
        <div>
            <input data-testid="search-input" type="text" placeholder="Search any shows here" value={searchEntry} onChange={handleChange}/>
        </div>
    );
};

export default SearchBox;