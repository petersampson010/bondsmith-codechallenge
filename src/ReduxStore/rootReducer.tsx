import { Reducer } from '@reduxjs/toolkit'

export const initialState = {
    premShows: [],
    searchedShows: [],
    searchActive: false,
    showDetailActive: false,
    activeShow: null
}

const rootReducer: Reducer<{}> = (state=initialState, action) => {
    switch(action.type) {
        case "UPDATEPREMSHOWS":
            return {
                ...state, 
                premShows: action.premShows
            }
        case "UPDATESHOWDETAILACTIVE":
            return {
                ...state, 
                showDetailActive: action.bool,
                activeShow: action.show
            }
        case 'UPDATESEARCHSHOWS':
            return {
                ...state, 
                searchedShows: action.shows,
                searchActive: true,
                showDetailActive: false,
                activeShow: null
            }
        default:
            return state;
    }
}

export default rootReducer;