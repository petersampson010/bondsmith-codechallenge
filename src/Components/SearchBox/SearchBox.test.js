import React from "react";
import { render, queryByTestId, fireEvent } from '@testing-library/react';
import SearchBox from "./SearchBox";
import { Provider } from 'react-redux';
import { legacy_createStore } from 'redux'
import rootReducer from '../../ReduxStore/rootReducer.tsx';
const store = legacy_createStore(rootReducer)

describe('ShowDetail Component Test', () => {
    test('matches snapshot', () => {
        const { container } = render(<Provider store={store}><SearchBox/></Provider>);
        expect(container).toMatchSnapshot();
    });
    test('debounce functionality is working', () => {
        const { queryByTestId } = render(<Provider store={store}><SearchBox/></Provider>);
        const searchInput = queryByTestId('search-input');
        fireEvent.change(searchInput, {target: {value:  "Batman"}});
        expect(window.location.search).toEqual("");
    });
    test('URL updates with search input', async() => {
        const { queryByTestId } = render(<Provider store={store}><SearchBox/></Provider>);
        const searchInput = queryByTestId('search-input');
        fireEvent.change(searchInput, {target: {value:  "Batman"}});
        await new Promise((r) => setTimeout(r, 2000));
        expect(window.location.search.split('=')[1]).toEqual("Batman");
    });
})