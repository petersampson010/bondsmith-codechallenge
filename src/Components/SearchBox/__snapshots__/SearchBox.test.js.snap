// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowDetail Component Test matches snapshot 1`] = `
<div>
  <div>
    <input
      data-testid="search-input"
      placeholder="Search any shows here"
      type="text"
      value=""
    />
  </div>
</div>
`;
